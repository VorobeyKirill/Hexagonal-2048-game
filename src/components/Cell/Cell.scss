@import "../../variables.scss";

.cell {
    position: absolute;

    background: $primary-color;
    clip-path: polygon(25% 0%, 75% 0%, 100% 50%, 75% 100%, 25% 100%, 0% 50%);
}

.cell__tile {
    position: absolute;

    background: $tile-color-0;
    clip-path: polygon(25% 0%, 75% 0%, 100% 50%, 75% 100%, 25% 100%, 0% 50%);

    font-weight: 400;

    display: flex;
    justify-content: center;
    align-items: center;

    &[data-cell-value="2"] {
        background: $tile-color-2;
    }

    &[data-cell-value="4"] {
        background: $tile-color-4;
    }

    &[data-cell-value="8"] {
        background: $tile-color-8;
    }

    &[data-cell-value="16"] {
        background: $tile-color-16;
    }

    &[data-cell-value="32"] {
        background: $tile-color-32;
    }

    &[data-cell-value="64"] {
        background: $tile-color-64;
    }

    &[data-cell-value="128"] {
        background: $tile-color-128;
    }

    &[data-cell-value="256"] {
        background: $tile-color-256;
    }

    &[data-cell-value="512"] {
        background: $tile-color-512;
    }

    &[data-cell-value="1024"] {
        background: $tile-color-1024;
    }

    &[data-cell-value="2048"] {
        background: $tile-color-2048;
    }
}