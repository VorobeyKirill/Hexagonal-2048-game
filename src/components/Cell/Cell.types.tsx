export interface CellProps {
    cellData: {
        x: number;
        y: number;
        z: number;
        value: number;
    },
    size: number;
    boardRadius: number;
}