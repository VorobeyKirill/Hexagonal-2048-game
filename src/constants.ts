export const BOARD_WIDTH = 500;