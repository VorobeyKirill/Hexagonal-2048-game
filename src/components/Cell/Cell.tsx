import { FC } from "react";
import { CELL_BORDER_WIDTH } from "../../constants";

import "./Cell.scss";
import { useScale } from "../../hooks/useScale";
import { CellProps } from "./Cell.types";

export const Cell: FC<CellProps> = ({ cellData, size, boardRadius }) => {
    const {x, y, z, value } = cellData;

    const scaleValue = useScale(value);

    const cellWidth = size * 2;
    const cellHeight = size * Math.sqrt(3);

    const leftOffset = size * 1.5 * x + Math.sqrt(3) * (boardRadius - 1) * cellHeight / 2 - CELL_BORDER_WIDTH * x;
    const topOffset = -cellHeight / 2 * (1 + x + 2 * y) + CELL_BORDER_WIDTH * y + CELL_BORDER_WIDTH * x / Math.sqrt(3) + boardRadius * cellHeight * 1.5 / Math.sqrt(3);

    return (
        <div
            style={{
                width: `${cellWidth}px`,
                height: `${cellHeight}px`,
                top: `${topOffset}px`,
                left: `${leftOffset}px`
            }}
            className="cell"
            data-x={x}
            data-y={y}
            data-z={z}
            data-value={value}
            data-testid="cell"
        >
            <div
                className="cell__tile"
                data-testid="cell-tile"
                style={{
                    width: `${cellWidth - CELL_BORDER_WIDTH * 4 + boardRadius}px`,
                    height: `${cellHeight - CELL_BORDER_WIDTH * 4 + boardRadius}px`,
                    top: `${CELL_BORDER_WIDTH * 2 - boardRadius / 2}px`,
                    left: `${CELL_BORDER_WIDTH * 2 - boardRadius / 2}px`,
                    transform: `scale(${scaleValue})`
                }}
                data-cell-value={value}
            >
                {!!value && <span>{value}</span>}
            </div>
        </div>
    );
}