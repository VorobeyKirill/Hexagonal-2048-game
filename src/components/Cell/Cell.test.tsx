import { render, screen } from "@testing-library/react";
import { Cell } from "./Cell";
import { CELL_BORDER_WIDTH } from "../../constants";

describe("Cell Component", () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  test("matches snapshot", () => {
    const cellData = { x: 0, y: 0, z: 0, value: 2 };
    const size = 50;
    const boardRadius = 2;

    expect(
      render(<Cell cellData={cellData} size={size} boardRadius={boardRadius} />)
    ).toMatchSnapshot();
  });

  test("renders cell with correct coordinates and value", () => {
    const cellData = { x: 0, y: 0, z: 0, value: 2 };
    const size = 50;
    const boardRadius = 2;

    render(<Cell cellData={cellData} size={size} boardRadius={boardRadius} />);

    const cellElement = screen.getByTestId("cell");

    expect(cellElement).toBeInTheDocument();
    expect(cellElement).toHaveStyle({
      width: `${size * 2}px`,
      height: `${size * Math.sqrt(3)}px`,
    });
  });

  test("renders tile inside cell with correct styles and value", () => {
    const cellData = { x: 0, y: 0, z: 0, value: 2 };
    const size = 50;
    const boardRadius = 2;

    render(<Cell cellData={cellData} size={size} boardRadius={boardRadius} />);

    const tileElement = screen.getByTestId("cell-tile");

    expect(tileElement).toBeInTheDocument();
    expect(tileElement).toHaveStyle({
      width: `${size * 2 - CELL_BORDER_WIDTH * 4 + boardRadius}px`,
      height: `${size * Math.sqrt(3) - CELL_BORDER_WIDTH * 4 + boardRadius}px`,
      top: `${CELL_BORDER_WIDTH * 2 - boardRadius / 2}px`,
      left: `${CELL_BORDER_WIDTH * 2 - boardRadius / 2}px`,
    });
  });

  test("renders span with correct value inside tile", () => {
    const cellData = { x: 0, y: 0, z: 0, value: 2 };
    const size = 50;
    const boardRadius = 2;

    render(<Cell cellData={cellData} size={size} boardRadius={boardRadius} />);

    const spanElement = screen.getByText("2");

    expect(spanElement).toBeInTheDocument();
  });

  test("renders cell without value", () => {
    const cellData = { x: 0, y: 0, z: 0, value: 0 };
    const size = 50;
    const boardRadius = 2;

    render(<Cell cellData={cellData} size={size} boardRadius={boardRadius} />);

    const spanElement = screen.queryByText("0");

    expect(spanElement).toBeNull();
  });
});
