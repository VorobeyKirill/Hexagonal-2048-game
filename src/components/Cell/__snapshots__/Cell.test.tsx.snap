// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cell Component matches snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="cell"
        data-testid="cell"
        data-value="2"
        data-x="0"
        data-y="0"
        data-z="0"
        style="width: 100px; height: 86.60254037844386px; top: 106.69872981077808px; left: 75px;"
      >
        <div
          class="cell__tile"
          data-cell-value="2"
          data-testid="cell-tile"
          style="width: 82px; height: 68.60254037844386px; top: 9px; left: 9px; transform: scale(1.1);"
        >
          <span>
            2
          </span>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="cell"
      data-testid="cell"
      data-value="2"
      data-x="0"
      data-y="0"
      data-z="0"
      style="width: 100px; height: 86.60254037844386px; top: 106.69872981077808px; left: 75px;"
    >
      <div
        class="cell__tile"
        data-cell-value="2"
        data-testid="cell-tile"
        style="width: 82px; height: 68.60254037844386px; top: 9px; left: 9px; transform: scale(1.1);"
      >
        <span>
          2
        </span>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
